### ISSUE TITLE
Serverless framework integration

### ISSUE THREAD

**Issue by marktani** [graphcool/graphcool-framework#243]
<a href="https://github.com/schickling"><img src="https://avatars2.githubusercontent.com/u/1567498?v=4" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [schickling](https://github.com/schickling)**
_Friday Sep 29, 2017 at 08:46 GMT_
_Originally opened as https://github.com/graphcool/prisma/issues/691_

----

There are two ways to integrate with the [Serverless framework](https://github.com/serverless/serverless):

1) Reference Serverless functions/resources from `graphcool.yml`
2) Reference a Graphcool backend from `serverless.yml`

### Goal

Use the Serverless framework to deploy functions on AWS (or other providers) and reference the deployed functions in the `graphcool.yml` file.

### Design decisions

* Graphcool assumes that during deploy-time all serverless functions are already deployed and can be referenced directly as AWS Lambda function

### Implementation

_TBD_

### Challenges

* Avoid code duplication when referencing a `serverless.yml` file
* IAM permissions (role statements)
* Zero-downtime deployment (blue-green)

### Initial limitations

* Supports only AWS Lambda (other service providers can be added later)

### External todos

- [ ] Serverless needs to provide information about deployed functions/resources in a readable format (e.g. JSON)

## 2) Reference a Graphcool backend from `serverless.yml`

_TBD_

**Comment by marktani** [comment]
<a href="https://github.com/kbrandwijk"><img src="https://avatars1.githubusercontent.com/u/852069?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [kbrandwijk](https://github.com/kbrandwijk)**
_Friday Sep 29, 2017 at 09:20 GMT_

----

Initial thoughts:
- Having two different configuration files for two different parts of the same solution is really not optimal
- For 'basic' uses (inline functions and existing webhooks) the graphcool.yml method is easiest

- For advanced uses (especially now the CLI has been completely refactored this should be easy), think about exposing the CLI functionality as serverless plugin, and define the entire deployment in the serverless.yml.
- If you go that route, you are not limited to AWS Lambda, expecially now Serverless has just released their plugin for Webtask supporting Node 8!

**Comment by marktani** [comment]
<a href="https://github.com/kbrandwijk"><img src="https://avatars1.githubusercontent.com/u/852069?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [kbrandwijk](https://github.com/kbrandwijk)**
_Friday Sep 29, 2017 at 13:38 GMT_

----

I'll address a few different topics in different comments.

## Deploying using serverless to a single provider

With serverless, the provider determines where your functions will be deployed. So for deploying functions to Graphcool (as inline functions), Graphcool needs to be defined as provider. For deploying your functions to another provider, Graphcool can be defined as a plugin. For example:

```yaml
# serverless.yml with webtasks provider

service: myProject

provider:
  name: webtasks

functions:
  hello:
    handler: hello
    events:
    - graphcool:
       type: resolver
       schema: hello.graphql
  after-post-create:
    handler: after-post-create
    events:
    - graphcool:
        type: operationAfter
        operation: Post.create

custom:
  graphcool:
    permissions:
    - operation: Post.create
      authenticated: true

plugins:
  - '@webtask/serverless-webtasks'
  - 'serverless-graphcool'
```

```yaml
# serverless.yml with aws provider

service: myProject

provider:
  name: aws

functions:
  hello:
    handler: handlers.hello
    events:
    - http:
        path: hello
        method: get
    - graphcool:
        type: resolver
        schema: hello.graphql

custom:
  graphcool:
    permissions:
    - operation: Post.create
      authenticated: true

plugins:
  - 'serverless-graphcool'
```

```yaml
# serverless.yml with Graphcool provider

service: myProject

provider:
  name: graphcool

functions:
  hello:
    handler: hello
    events:
      - graphcool:
          type: resolver
          schema: hello.graphql

custom:
  graphcool:
    permissions:
    - operation: Post.create
      authenticated: true
```
Basically, anything that's in the graphcool.yml can be moved 1-on-1 under the custom/graphcool node in serverless.yml, except for the functions part, because that moves to the main serverless functions node.

**Comment by marktani** [comment]
<a href="https://github.com/kbrandwijk"><img src="https://avatars1.githubusercontent.com/u/852069?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [kbrandwijk](https://github.com/kbrandwijk)**
_Friday Sep 29, 2017 at 14:00 GMT_

----

## Mix-n-match

There might be scenarios, where you want to combine different services together, deployed to different providers:

```yaml
# service1/serverless.yml

service: service1

provider:
  name: webtasks

functions:
  hello:
    handler: hello
  after-post-create:
    handler: after-post-create

plugins:
  - '@webtask/serverless-webtasks'

# service2/serverless.yml

service: service2

provider:
  name: aws

functions:
  create-post:
    handler: handlers.createPost
    events:
    - http:
        path: create-post
        method: post

# serverless.yml

service: myProject

provider:
  name: graphcool

functions:
  hello:
    webhook: `${self:custom.webtaskBaseUrl}/hello`
    - events:
       - graphcool:
           type: resolver
           schema: hello.graphql
  create-post:
    webhook: `${self:custom.webtaskBaseUrl}/post-create`
    - events:
       - graphcool:
           type: operationAfter
           operation: Post.create

custom:
  webtaskBaseUrl: 'https://sandbox.it-auth0.com/container'
  awsBaseUrl: 'https://..........'
  graphcool:
    permissions:
    - operation: Post.create
      authenticated: true
```

**Comment by marktani** [comment]
<a href="https://github.com/kbrandwijk"><img src="https://avatars1.githubusercontent.com/u/852069?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [kbrandwijk](https://github.com/kbrandwijk)**
_Friday Sep 29, 2017 at 14:17 GMT_

----

>External todos
>
 >Serverless needs to provide information about deployed functions/resources in a readable format (e.g. JSON)

Serverless actually does provide this information to plugins. It depends which provider you use. Have a look at the info command for serverless-webtasks for example, which provides you with the endpoints. AWS has something similar: https://github.com/auth0/serverless-webtasks/blob/master/lib/info/service.js#L36-L51

**Comment by marktani** [comment]
<a href="https://github.com/kbrandwijk"><img src="https://avatars1.githubusercontent.com/u/852069?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [kbrandwijk](https://github.com/kbrandwijk)**
_Saturday Sep 30, 2017 at 18:30 GMT_

----

I also created https://github.com/graphcool/graphcool/issues/710, because that aligns better with the serverless concept of having different services, each in their own folder, with their own serverless.yml file. The only wait to align that with Graphcool, is to allow deployment of different serverless services to the same Graphcool project, essentially merging.

**Comment by marktani** [comment]
<a href="https://github.com/kbrandwijk"><img src="https://avatars1.githubusercontent.com/u/852069?v=4" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [kbrandwijk](https://github.com/kbrandwijk)**
_Monday Oct 23, 2017 at 12:25 GMT_

----

@schickling Update: I took a different direction for now for integrating Graphcool and Serverless. I have described some of the details here: https://gist.github.com/kbrandwijk/86683ad58a98447b0ddc5f7f11cdb789.
